File: kotlin/src/test/kotlin/web/snapshots/FileSystemPathProviderTest.kt
package web.snapshots

import org.junit.jupiter.api.DynamicTest.dynamicTest
import org.junit.jupiter.api.TestFactory
import web.url.FreeUrl
import kotlin.test.assertTrue

class FileSystemPathProviderTest {
    private val subject = FileSystemPathProvider()

    @TestFactory
    fun getSnapshotDirPath() = mapOf(
        "https://www.tumblr.com/getfursuit" to "T/tumblr.com/getfursuit",
        "https://furries.club/@getfursuit" to "F/furries.club/getfursuit",
        "https://github.com/veelkoov/fuzzrake/pull/194" to "G/github.com/veelkoov_fuzzrake_pull_194",
        "https://getfursu.it/data_updates.html#anchor" to "G/getfursu.it/data_updates.html",
    ).map { (url, expectedPrefix) ->
        dynamicTest(url) {
            val result = subject.getSnapshotDirPath(FreeUrl(url))

            assertTrue(result.startsWith(expectedPrefix), "Unexpected prefix: $result")
        }
    }
}

File: symfony/tests/Utils/Web/Snapshots/FileSystemPathProviderTest.php
<?php

declare(strict_types=1);

namespace App\Tests\Utils\Web\Snapshots;

use App\Utils\Web\Snapshots\FileSystemPathProvider;
use App\Utils\Web\Url\FreeUrl;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\Attributes\Small;
use PHPUnit\Framework\TestCase;

#[Small]
class FileSystemPathProviderTest extends TestCase
{
    /**
     * @return list<array{string, non-empty-string}>
     */
    public static function getSnapshotDirPathDataProvider(): array
    {
        return [
            ['https://www.tumblr.com/getfursuit', 'T/tumblr.com/getfursuit'],
            ['https://furries.club/@getfursuit', 'F/furries.club/getfursuit'],
            ['https://github.com/veelkoov/fuzzrake/pull/194', 'G/github.com/veelkoov_fuzzrake_pull_194'],
            ['https://getfursu.it/data_updates.html#anchor', 'G/getfursu.it/data_updates.html'],
        ];
    }

    /**
     * @param non-empty-string $expectedPrefix
     */
    #[DataProvider('getSnapshotDirPathDataProvider')]
    public function testGetSnapshotDirPath(string $inputUrl, string $expectedPrefix): void
    {
        $subject = new FileSystemPathProvider();

        self::assertStringStartsWith($expectedPrefix, $subject->getSnapshotDirPath(new FreeUrl($inputUrl, '')));
    }
}
